### DISCUSSION TITLE
Need mechanism to ensure that code updates do not break case studies

### DISCUSSION THREAD

**Discussion by iincer** [pacti-org/pacti#334] [General]
Recent features under discussion can include incompatibilities in the API. As Pacti is updated, we want to make sure that the case studies, which are reference code, are updated, too. One way to do this would be to have means of executing all notebooks from a centralized location.
How do we execute the notebooks as python files while getting the feedback it worked/didn't work?

**Comment by murrayrm** [comment]
Here's a script that will run notebooks and report on whether there were errors (can be included in GitHub Actions):
```
#!/bin/bash
# run_notbooks.sh - run Jupyter notebooks
# RMM, 26 Jan 2021
#
# This script runs all of the Jupyter notebooks to make sure that there
# are no errors.

# Keep track of files that generate errors
notebook_errors=""

# Go through each Jupyter notebook
for example in *.ipynb; do
    echo "Running ${example}"
    if ! jupyter nbconvert --to notebook --execute ${example}; then
        notebook_errors="${notebook_errors} ${example}"
    fi
done

# Get rid of the output files
rm *.nbconvert.ipynb

# List any files that generated errors
if [ -n "${notebook_errors}" ]; then
    echo These examples had errors:
    echo "${notebook_errors}"
    exit 1
else
    echo All examples ran successfully
fi
```

**Comment by iincer** [comment]
Perfect. Thanks, @murrayrm!!
